File: Program's_Contributed_By_Contributors/C++ Programs/chocolate_distribution.cpp
//Chocolate Distribution problem


/*  Given an array A[ ] of positive integers of size N, where each value represents the number of chocolates in a packet. Each packet can have a variable number of chocolates. There are M students, the task is to distribute chocolate packets among M students such that :
1. Each student gets exactly one packet.
2. The difference between maximum number of chocolates given to a student and minimum number of chocolates given to a student is minimum.   */


#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
    long long findMinDiff(vector<long long> a, long long n, long long m){
    
    sort(a.begin(),a.end());
        long long i = 0, j = 0,difference;
        difference = INT_MAX;
        while(j < n)
        {
            if(j-i+1 < m)
                j++;
            else if(j-i+1 == m)
            {
                difference = min(difference, a[j]-a[i]);
                i++;
                j++;
            }
        }
        return difference;
}
};

int main() {
	long long t;
	cin>>t;
	while(t--)
	{
		long long n;
		cin>>n;
		vector<long long> a;
		long long x;
		for(long long i=0;i<n;i++)
		{
			cin>>x;
			a.push_back(x);
		}
		
		long long m;
		cin>>m;
		Solution ob;
		cout<<ob.findMinDiff(a,n,m)<<endl;
	}
	return 0;
}

File: Program's_Contributed_By_Contributors/Python_Programs/Chocolate Distribution.py
# Python3 program to solve chocolate
# distribution problem

# arr[0..n-1] represents sizes of packets
# m is number of students.
# Returns minimum difference between maximum
# and minimum values of distribution.
def findMinDiff(arr, n, m):

	# if there are no chocolates or number
	# of students is 0
	if (m==0 or n==0):
		return 0

	# Sort the given packets
	arr.sort()

	# Number of students cannot be more 
	# than number of packets
	if (n < m):
		return -1

	# Largest number of chocolates
	min_diff = arr[n-1] - arr[0]

	# Find the subarray of size m such that
	# difference between last (maximum in case
	# of sorted) and first (minimum in case of
	# sorted) elements of subarray is minimum.
	for i in range(len(arr) - m + 1):
		min_diff = min(min_diff , 
					arr[i + m - 1] - arr[i]) 
		
	return min_diff

# Driver Code
if __name__ == "__main__":
	arr = [12, 4, 7, 9, 2, 23, 25, 41,
		30, 40, 28, 42, 30, 44, 48, 
		43, 50]

	# Number of students
	m = 7
	n = len(arr)
	print("Minimum difference is", 
		findMinDiff(arr, n, m))
# This code is contributed by Smitha
